consensus/consensus_v2.go

```diff
@@ -100,60 +102,70 @@ func (consensus *Consensus) HandleMessageUpdate(ctx context.Context, msg *msg_pb
 	return nil
 }
 
-func (consensus *Consensus) finalizeCommits() {
+func (consensus *Consensus) finalCommit() {
 	consensus.getLogger().Info().
 		Int64("NumCommits", consensus.Decider.SignersCount(quorum.Commit)).
-		Msg("[finalizeCommits] Finalizing Block")
+		Msg("[finalCommit] Finalizing Consensus")
 	beforeCatchupNum := consensus.blockNum
+
 	leaderPriKey, err := consensus.GetConsensusLeaderPrivateKey()
 	if err != nil {
-		consensus.getLogger().Error().Err(err).Msg("[FinalizeCommits] leader not found")
+		consensus.getLogger().Error().Err(err).Msg("[finalCommit] leader not found")
 		return
 	}
 	// Construct committed message
 	network, err := consensus.construct(msg_pb.MessageType_COMMITTED, nil, []*bls.PrivateKeyWrapper{leaderPriKey})
 	if err != nil {
 		consensus.getLogger().Warn().Err(err).
-			Msg("[FinalizeCommits] Unable to construct Committed message")
+			Msg("[finalCommit] Unable to construct Committed message")
 		return
 	}
-	msgToSend, aggSig, FBFTMsg :=
+	msgToSend, FBFTMsg :=
 		network.Bytes,
-		network.OptionalAggregateSignature,
 		network.FBFTMsg
-	consensus.aggregatedCommitSig = aggSig // this may not needed
+	commitSigAndBitmap := FBFTMsg.Payload
 	consensus.FBFTLog.AddMessage(FBFTMsg)
 	// find correct block content
 	curBlockHash := consensus.blockHash
 	block := consensus.FBFTLog.GetBlockByHash(curBlockHash)
 	if block == nil {
 		consensus.getLogger().Warn().
 			Str("blockHash", hex.EncodeToString(curBlockHash[:])).
-			Msg("[FinalizeCommits] Cannot find block by hash")
+			Msg("[finalCommit] Cannot find block by hash")
 		return
 	}
 
-	consensus.tryCatchup()
-	if consensus.blockNum-beforeCatchupNum != 1 {
-		consensus.getLogger().Warn().
+	consensus.getLogger().Info().Hex("new", commitSigAndBitmap).Msg("[finalCommit] Overriding commit signatures!!")
+	consensus.Blockchain.WriteCommitSig(block.NumberU64(), commitSigAndBitmap)
+
+	block.SetCurrentCommitSig(commitSigAndBitmap)
+	err = consensus.commitBlock(block, FBFTMsg)
+
+	if err != nil || consensus.blockNum-beforeCatchupNum != 1 {
+		consensus.getLogger().Err(err).
 			Uint64("beforeCatchupBlockNum", beforeCatchupNum).
-			Msg("[FinalizeCommits] Leader cannot provide the correct block for committed message")
+			Msg("[finalCommit] Leader failed to commit the confirmed block")
 		return
 	}
 
-	// if leader success finalize the block, send committed message to validators
+	// if leader successfully finalizes the block, send committed message to validators
+	sendImmediately := false
+	if !consensus.IsLeader() || block.IsLastBlockInEpoch() {
```

Comment:
finalCommit only happens on leader. So leader will send 67% commit immediately. the 100% commit won't be sent immediately to save network traffic. It will only be sent in retry if consensus doesn't move forward. Or if the leader is changed, so the 100% committed should be sent to the next leader immediately.